cpp/src/io/parquet/decode_fixed.cu

```diff
@@ -1252,15 +1190,44 @@ CUDF_KERNEL void __launch_bounds__(decode_block_size_t, 8)
     }
 
     // decode the values themselves
-    if constexpr (has_strings_t) {
-      string_output_offset = decode_strings<decode_block_size_t, has_lists_t, split_decode_t>(
-        s, sb, valid_count, next_valid_count, t, string_output_offset);
-    } else if constexpr (split_decode_t) {
-      decode_fixed_width_split_values<decode_block_size_t, has_lists_t>(
-        s, sb, valid_count, next_valid_count, t);
+    if constexpr (has_lists_t) {
+      if constexpr (has_strings_t) {
+        string_output_offset =
+          decode_strings<decode_block_size_t, has_lists_t, split_decode_t, copy_mode::INDIRECT>(
+            s, sb, valid_count, next_valid_count, t, string_output_offset);
+      } else if constexpr (split_decode_t) {
+        decode_fixed_width_split_values<decode_block_size_t, has_lists_t, copy_mode::INDIRECT>(
+          s, sb, valid_count, next_valid_count, t);
+      } else {
+        decode_fixed_width_values<decode_block_size_t, has_lists_t, copy_mode::INDIRECT>(
+          s, sb, valid_count, next_valid_count, t);
+      }
     } else {
-      decode_fixed_width_values<decode_block_size_t, has_lists_t>(
-        s, sb, valid_count, next_valid_count, t);
+      if (!should_process_nulls) {
+        if constexpr (has_strings_t) {
+          string_output_offset =
+            decode_strings<decode_block_size_t, has_lists_t, split_decode_t, copy_mode::DIRECT>(
+              s, sb, valid_count, next_valid_count, t, string_output_offset);
+        } else if constexpr (split_decode_t) {
+          decode_fixed_width_split_values<decode_block_size_t, has_lists_t, copy_mode::DIRECT>(
+            s, sb, valid_count, next_valid_count, t);
+        } else {
+          decode_fixed_width_values<decode_block_size_t, has_lists_t, copy_mode::DIRECT>(
+            s, sb, valid_count, next_valid_count, t);
+        }
+      } else {
+        if constexpr (has_strings_t) {
+          string_output_offset =
+            decode_strings<decode_block_size_t, has_lists_t, split_decode_t, copy_mode::INDIRECT>(
+              s, sb, valid_count, next_valid_count, t, string_output_offset);
+        } else if constexpr (split_decode_t) {
+          decode_fixed_width_split_values<decode_block_size_t, has_lists_t, copy_mode::INDIRECT>(
+            s, sb, valid_count, next_valid_count, t);
+        } else {
+          decode_fixed_width_values<decode_block_size_t, has_lists_t, copy_mode::INDIRECT>(
+            s, sb, valid_count, next_valid_count, t);
+        }
+      }
     }
```

Comment:
Can we simplify this to:

```cpp
if (has_strings_t or should_process_nulls)
{
  // use copy_mode::INDIRECT
}
else
{
  // use copy_mode::DIRECT
}
```
It looks quite involved and prone to errors in the future this way

Reply:
That's fine but does changing this block to the suggested one (not use `has_strings_t` as constexpr) make a performance impact?